[graphql] Resolver to terminate runs

clairelin135 opened the pull request:
This PR adds a resolver to terminate a set of runs, intended to support a "terminate all queued runs" button. Under the hood, it calls `instance.optimize_for_webserver` to pool connections and reduce query time. The implementation in this PR takes 3.5 min to terminate 5K runs, compared to ~15 min without the `optimize_for_webserver` call.

The resolver accepts a list of run IDs. This is intended so that the front end can call `instance.get_run_ids(cursor=..., limit=...)` to fetch run IDs in batches, then pass this to the resolver. This enables displaying a progress bar.

gibsondan commented on python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py:
```diff
@@ -143,6 +151,41 @@ def terminate_pipeline_execution(
+                graphene_info,
+                instance,
+                run_id,
+                cast(GrapheneTerminateRunPolicy, GrapheneTerminateRunPolicy.SAFE_TERMINATE),
```
it seems to me like for this use case where you want lots of runs terminated ASAP we want the one that moves it into canceled immediately (GrapheneTerminateRunPolicy.MARK_AS_CANCELED_IMMEDIATELY)

or you could pass it in as an argument

gibsondan commented on python_modules/dagster-graphql/dagster_graphql/schema/roots/mutation.py:
```diff
@@ -249,6 +250,27 @@ class Meta:
+class GrapheneTerminateRunsResult(graphene.ObjectType):
+    """Indicates the runs that successfully terminated and those that failed to terminate."""
+
+    runIdsTerminated = non_null_list(graphene.String)
+    runIdsFailedToTerminate = non_null_list(graphene.String)
```
this could also be a list of GrapheneTerminateRunFailure and a list of GrapheneTerminateRunSuccess? that way it would include the failure message if there is one

clairelin135 replied:
I'll return a list of `GrapheneTerminateRunFailure`s, agree the error message would be helpful.

I think we can continue returning the run ids successfully terminated though, `GrapheneTerminateRunSuccess` returns the whole run object which could be a lot for tens of thousands of runs

gibsondan replied:
it's graphql so it only returns the fields they request right?

i.e. you could do 
```
{
   ... on TerminateRunSuccess {
     run {
       id
     }
   }
}
```

and only get the run IDs back

clairelin135 replied:
ah right good call

gibsondan reviewed (changes requested):
mostly small things inline, broadly makes sense to me

clairelin135 commented:
Added the termination option as an arg after talking to @bengotow, in order to enable the UI to use this mutation outside of the "cancel all queued runs" query

clairelin135 commented on python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py:
```diff
@@ -91,21 +96,27 @@ def terminate_pipeline_execution(
-    if not record:
-        assert_permission(graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION)
-        return GrapheneRunNotFoundError(run_id)
+    try:
```
wrap the permissions assertions in a try/except so that we can return an error containing a run ID. 

This also allows executing the `terminateRuns` mutation without exiting on the first permissions error

gibsondan commented on python_modules/dagster-graphql/dagster_graphql/schema/roots/mutation.py:
```diff
@@ -235,6 +236,19 @@ class Meta:
         name = "TerminateRunFailure"
 
 
+class GrapheneTerminateRunUnauthorizedError(graphene.ObjectType):
```
Is it useful to have both this and GrapheneTerminateRunFailure? If the frontend doesn't treat them differently, it seems like you could remove this class and use that instead (and the reason it failed is b/c you're unauthorized)

gibsondan commented on python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py:
```diff
@@ -91,21 +96,27 @@ def terminate_pipeline_execution(
+        else:
+            assert_permission(graphene_info, Permissions.TERMINATE_PIPELINE_EXECUTION)
+
+    except UserFacingGraphQLError as e:
+        return GrapheneTerminateRunUnauthorizedError(
+            run_id=run_id, message=e.error.message if e.error.message else "Authorization failed"
         )
```
i don't think its necc correct to assume that any UserFacingGraphQLError is a permissions error (plus i think its good to avoid using exceptions as control flow)

What i would suggest is removing the try/catch, and instead of calling assert_permission and assert_permission_for_location and potentially raising an exception, check has_permission and has_permission_for_location, and return the value you want if they fail

I think this could return a GrapheneTerminateRunFailure instead of a new type that the frontend needs to worry about - do we need to worry at all about this affecting the frontend callsite for the single run termination graphql call?

gibsondan commented on python_modules/dagster-graphql/dagster_graphql/schema/roots/mutation.py:
```diff
@@ -243,12 +257,32 @@ class Meta:
             GrapheneTerminateRunFailure,
             GrapheneRunNotFoundError,
-            GrapheneUnauthorizedError,
+            GrapheneTerminateRunUnauthorizedError,
```
if we add this new type, we should be sure that the frontend knows how to handle it

gibsondan reviewed (approved):
accepting to unblock but I would suggest removing TerminateRunUnauthorizedError and using TerminateRunFailure instead (unless the frontend is specifically handling the unauthorized case - but if it was, we would need to change it as part of this PR too I think right?)

Couple other things inline wrt how we handle the permission failure case

clairelin135 commented:
@gibsondan thanks for taking a look -- I made the following changes based on what you mentioned:
- Removed the try/except in favor of returning `TerminateRunFailure`
- Added a `canBulkTerminate` resolver to check if you have deployment-wide permissions

I ended up implementing both because we may replace the existing `terminateRun` resolver to just call this new mutation, and it would be ideal to receive a per-run exception rather than raise an exception that blocks all subsequent execution
